Refactor expression runner so it can be used via the C and JS APIs

dcodeIO opened the pull request:
Refactors most of the precompute pass's expression runner into its base class so it can also be used via the C and JS APIs. Also adds the option to populate the runner with known constant local and global values upfront, and remembers assigned intermediate values as well as traversing into functions if requested.

### C-API:

* **ExpressionRunnerFlagsDefault**()
* **ExpressionRunnerFlagsPreserveSideeffects**()
* **ExpressionRunnerFlagsTraverseCalls**()
* **ExpressionRunnerCreate**(module, flags, maxDepth, maxLoopIterations)
* **ExpressionRunnerSetLocalValue**(runner, index, value)
* **ExpressionRunnerSetGlobalValue**(runner, name, value)
* **ExpressionRunnerRunAndDispose**(runner, expr)

### JS-API:

* binaryen.**ExpressionRunner.Flags.Default**
* binaryen.**ExpressionRunner.Flags.PreserveSideeffects**
* binaryen.**ExpressionRunner.Flags.TraverseCalls**
* new binaryen.**ExpressionRunner**(module[, flags[, maxDepth[, maxLoopIterations]]])
* binaryen.**ExpressionRunner#setLocalValue**(index, valueExpr)
* binaryen.**ExpressionRunner#setGlobalValue**(name, valueExpr)
* binaryen.**ExpressionRunner#runAndDispose**(expr)

tlively commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2228,107 @@ class ModuleInstance
+  // 7 through it). When we want to replace the expression, we can only do so
+  // when it has no side effects. When we don't care about replacing the
+  // expression, we just want to know if it will contain a known constant.
+  bool replaceExpression;
```
This seems semantically complex enough that having an enum to name the options would be helpful.

dcodeIO commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2228,114 @@ class ModuleInstance
+    : ExpressionRunner<StandaloneExpressionRunner>(maxDepth), module(module),
+      getValues(getValues), intent(intent) {}
+
+  virtual ~StandaloneExpressionRunner() {}
```
I don't know what I'm doing here btw, but this was necessary to silence an error when `delete`ing the runner again. That correct?

tlively replied:
Hmm, this is a little fishy because there is no inheritance or any other virtual methods here. What is the error you get if you delete this line?

dcodeIO replied:
The error is

```
binaryen-c.cpp:4864:10: error: deleting object of polymorphic class type
‘wasm::StandaloneExpressionRunner’ which has non-virtual destructor might
cause undefined behavior [-Werror=delete-non-virtual-dtor]

   delete R;
          ^
```

tlively replied:
Ok, I figured out what was going on here. One of the parent classes of `ContextAwareExpressionRunner` has virtual methods, so the compiler pessimistically assumes that this class might have subclasses, so it requires a virtual destructor implementation. But if you change the declaration `class ContextAwareExpressionRunner` to `class ContextAwareExpressionRunner final`, you can tell the compiler that there will be no more subclasses and it will be able to use the default destructor. What a language.

dcodeIO replied:
Great, that worked! :)

dcodeIO commented:
Last commit adds a mechanism to support evaluating expressions like

```js
runner.runAndDispose(
  module.i32.add(
    module.block(null, [
      module.local.set(0, module.i32.const(5)),
      module.local.get(0, binaryen.i32)
    ], binaryen.i32),
    module.i32.const(1)
  )
);
```

where a value is `local.set` and later picked up again by a `local.get`, which is a common pattern in our code. Appears that precompute tests are unaffected by this, but chances are that I overlooked something. For instance, at one point I'm returning `Flow()` to indicate no value, instead of a breaking `Flow(NONSTANDALONE_FLOW)`, which I'm not certain about.

dcodeIO commented:
This now also implements what I mentioned earlier, in that constant local and global values known beforehand can be set explicitly. NFCI for existing code using the runner, but will open up new possibilities in that a generator does not have to inline these values manually, but can instead use locals as long as their constant value is known, e.g. `const` in JS where sometimes the value is known to be a compile time constant, and sometimes is just meant to be readonly.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2228,186 @@ class ModuleInstance
+  // replacing the expression, we just want to know if it will contain a known
+  // constant.
+  enum Intent {
+    // Intent is to evaluate the expression, so we can ignore some side effects
```
`some` seems to imply that certain side effects are ok but others are not, or that there is some amount of them that is ok. maybe remove that word?

dcodeIO replied:
The `some` was intended, in that for example a `local.tee` is ok, but branching or a trap of course is not. Perhaps I should make that more clear?

dcodeIO replied:
Made this more clear in the comments

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2228,186 @@ class ModuleInstance
     : ModuleInstanceBase(wasm, externalInterface) {}
 };
 
+static const Name NONSTANDALONE_FLOW("Binaryen|nonstandalone");
```
this const and the typedef after it are now in the general global scope, while before they were just in Precompute's global scope. Please move them into the new class.

dcodeIO replied:
I might be lacking C++ experience here, but iiuc moving the name to the class has the side-effect that it can't be static anymore, since it isn't an integer, so I ended up with one `const` definition inside the class and another `static const` definition of the same name in the precompute pass since it cannot reference the former, which has the potential to break if the respective string is changed at either place. Is there a better way?

tlively replied:
You can have a static member inside the class, but you have to initialize it outside the class with `type ClassName::member = initial_value`. Probably best to have a new .cpp file that holds this and the rest of the class implementation.

dcodeIO replied:
Thanks, I think I figured it out. Always a pleasure to learn some C++ from you guys, enjoying it :)

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2228,186 @@ class ModuleInstance
+  // map local indexes to values set in the context of this runner
+  std::unordered_map<Index, Literals> setLocalValues;
+  // map global names to values set in the context of this runner
+  std::unordered_map<Name, Literals> setGlobalValues;
```
IIUC, `getValues` is a mapping of specific gets to values, while `setLocalValues` maps indexes to values, so it applies to all gets for that index? That seems somewhat redundant.

One option might be to have just the former, and e.g. `visit()` could receive optional maps of `index => value` that it would then use to scan for gets and mark them with the proper value using the single mechanism.

Alternatively maybe we can rename/document this better so it's clearer.

dcodeIO replied:
From what I understood `getValues` is special in that it is referenced from the precompute pass and is possibly very large, since it covers all the gets up to the point where the runner was created. So, copying from there to a new map or vice-versa might involve quite some overhead. Going to see what can be done.

dcodeIO replied:
So it seems that no matter how we turn it, if we'd like to unify both concepts we'll end up polluting `getValues` of the precompute pass, except if we copy the whole map beforehand, which appears to be intentionally avoided because it can be large. What I did now is make `getValues` optional, as a pointer (which was necessary anyway for the C API), still taking precedence over manually set values as it's a pass-only mechanism (either or) anyway. Does that sound reasonable?

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2228,186 @@ class ModuleInstance
+  // only do so when it has no side effects. When we don't care about
+  // replacing the expression, we just want to know if it will contain a known
+  // constant.
+  enum Intent {
```
Thoughts on maybe `Mode` instead of `Intent`?

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2228,186 @@ class ModuleInstance
+
+// Evaluates a standalone expression. Errors if we hit anything that can't be
+// evaluated.
+class StandaloneExpressionRunner
```
I'm not sure about the name here. We evaluate a single expression, yes, but it isn't "standalone" in the sense that we have local and global values, and the globals in particular are from the "outside" in some sense...

What is mainly added here are local and global values, but I can't think of a good name for that atm...

dcodeIO replied:
Conceptionally it evaluates an expression "out of context of its function", except if one gives it context by specifying `getValues` or explicitly setting local or global values, while still assuming some global context (most notably immutable globals). Perhaps `GeneralizedExpressionRunner`?

dcodeIO commented:
Last commit now also traverses into (simple) functions, which has some minor but beneficial effects on existing tests. However, for this to work I had to add a `trapIfInvalid` flag to the base class since it would otherwise `abort` on some test cases like `test/unit.wat`, which seems odd and might indicate that there is something wrong either in my code or in general.

dcodeIO commented:
Hmm, appears this leads to non-determinism now, depending on the order that functions become optimized in, so sometimes a function is simple enough to evaluate and sometimes it's not (yet) :(

tlively commented:
Right, function-parallel passes like Precompute should not look at any functions outside the one they are currently optimizing. For sequences of function-parallel passes, rather than running one pass at a time on all the functions, the entire sequence is run separately on each function. This increases parallelism and cache locality, but means that different passes can be running on different functions concurrently.

dcodeIO commented:
I solved this by introducing two new modes `EVALUATE_DETERMINISTIC` and `REPLACE_DETERMINISTIC` which do not attempt to traverse calls (or in the future do anything else that'd be non-deterministic), so the precompute pass can stay function parallel. Makes me wonder though if it would be better to have `Flags` instead of a `Mode`, so one can set `REPLACE | DETERMINISTIC` which is a bit simpler?

tlively commented:
Yeah, if we're adding orthogonal options like that, a flags mechanism makes sense to me.

tlively commented on src/wasm/wasm-interpreter-ContextAwareExpressionRunner.cpp:
```diff
@@ -0,0 +1,276 @@
+/*
```
How about naming this file just `src/wasm/wasm-interpreter.cpp`.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2235,119 @@ class ModuleInstance
+
+  // Flags indicating special requirements, for example whether we are just
+  // evaluating (default), also going to replace the expression afterwards or
+  // executing in a function-parallel scenario.
```
this comment is duplicated above - one of the places would be enough I think.

dcodeIO replied:
Shortened the comment, referring to `FlagValues`. So we still have intellisense on it, but less duplicate text.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2235,119 @@ class ModuleInstance
+    // non-deterministically, like function calls, where the called function
+    // might or might not have been optimized already to something we can
+    // traverse successfully.
+    PARALLEL = 1 << 1
```
I think `PARALLEL` is not obvious enough. How about `TRAVERSE_OTHER_FUNCS` which directly says what is the underlying issue here?

dcodeIO replied:
Renamed the flags to `PRESERVE_SIDEEFFECTS` and `TRAVERSE_CALLS`

kripken commented on src/wasm/wasm-interpreter.cpp:
```diff
@@ -19,4 +35,252 @@ void Indenter::print() {
+  // traverse successfully.
+  if (!(flags & FlagValues::PARALLEL)) {
+    // Note that we are not a validator, so we skip calls to functions that do
+    // not exist yet or where the signature does not match.
```
This seems a little brittle - the compiler might end up emitting some nondeterministic. Is it worth handling these cases? I lean towards asserting that if we are allowed to traverse other functions, then they exist and are valid.

dcodeIO replied:
Makes sense, yeah. Change it, and doesn't hit any assertions anymore now that precompute doesn't use it due to parallelism.

kripken commented on src/wasm/wasm-interpreter.cpp:
```diff
@@ -19,4 +35,252 @@ void Indenter::print() {
+        auto numOperands = curr->operands.size();
+        if (numOperands == func->getNumParams()) {
+          ContextAwareExpressionRunner runner(module, flags, maxDepth);
+          runner.depth = depth + 1;
```
could be `runner.depth++`

dcodeIO replied:
Just checked, but this is essentially `other.depth = this.depth + 1` and should be fine :)

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2235,119 @@ class ModuleInstance
+  // Optional reference to a map of gets to constant values, for example where a
+  // pass did already compute these. Not applicable in C-API usage and possibly
+  // large, hence a reference to the respective map (must not be mutated).
+  GetValues* getValues;
```
can do ` = nullptr` here instead of in the class constructor

dcodeIO replied:
Removed the constructor prameter, initialized with `= nullptr` and made this a setter `setGetValues(...)` now, which only precompute uses.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -2228,6 +2235,119 @@ class ModuleInstance
+  Module* getModule();
+
+  // Sets a known local value to use. Returns `true` if the value is actually
+  // constant.
```
"actually constant" seems a little inaccurate. how about "is concrete (i.e., not unreachable)"

kripken commented on src/wasm/wasm-interpreter.cpp:
```diff
@@ -19,4 +35,252 @@ void Indenter::print() {
+      assert(curr->isTee());
+      return visit(curr->value);
+    }
+    // Otherwise remember the constant value set, if any, for subsequent gets.
```
Not sure I understand the model here. This seems to assume there is no control flow at all, i.e., nothing like
```
if (..) {
  x = 10;
}
f(x); // x might or might not be 10
```
Is that the assumption? How are we sure it's enforced?

In particular, the name of the class doesn't imply this only considers linear execution traces. If it were `LinearExpressionRunner` then this would be unsurprising.

Perhaps we do want separate enough things for a linear execution runner that it could derive from the context aware one?

dcodeIO replied:
This got me thinking. Iiuc the base `ExpressionRunner` is already linear, only traversing branches it knows are taken. Did I miss something there? If not, an alternative strategy might be to move all of the new runner into the base `ExpressionRunner` as its default implementation, with deriving runners overriding what they need to override, like one for precompute that uses `getValues`?

kripken replied:
Oh yes, after thinking on this, I think you're right. We should only be traversing code linearly, and halting if we hit anything that isn't linear anyhow. So yes, we should be able to always track locals that are set and read.

Ok, then yes, I think we can move that into the main runner, which will make this reasonably simple I hope.

kripken replied:
(and sorry for the delay here!)

dcodeIO commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1252,170 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+        if (func->sig.results.isConcrete()) {
+          auto numOperands = curr->operands.size();
+          assert(numOperands == func->getNumParams());
+          ExpressionRunner<SubType> runner(flags);
```
This is bad. Any suggestions how to model this correctly?

dcodeIO replied:
Fixed this by simply reusing the existing runner, which I should have been doing from the start :)

dcodeIO commented:
Alright, refactored all of this into the base `ExpressionRunner`. A couple things:

* There is [this unfortunate place](https://github.com/WebAssembly/binaryen/pull/2702#discussion_r401713557) where a runner wants to make a sub runner of the same kind, which, well, is bad. Open to any suggestions :)
* Since the base class is a template, I didn't attempt to move the implementation into its own source file so we don't have to deal with explicit template instantiation. Or should I?
* I initially attempted to add the concrete `CExpressionRunner` to `binaryen.cpp` to separate things a bit, but that didn't work due to an `empty struct has size 0 in C but 1 in CPP` error on the exception class. Is it ok if it lives in `wasm-interpreter.h` or should I use a workaround like a dummy field?
* Previously, the precompute runner would not attempt to traverse loops since these can be infinite. The base class, however, does attempt this. Now we don't know what a C-API user intends to do, so just overriding `visitLoop` in both precompute and the C runner didn't feel right. Ended up making it an option `maxLoopIterations` defaulting to `0` for no limit, that can be set from the C-API, and removed it along `maxDepth`, where `0` now also indicates no limit, from the constructor signature so the ctor doesn't become unnecessarily ugly. Might look like an unnecessary change at first, so I thought I explain why.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -170,11 +178,71 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+
+  Flags flags;
+
+  ExpressionRunner(Flags flags = FlagValues::DEFAULT) : flags(flags) {}
```
You can add another constructor that accepts `module, maxDepth, maxLoopIteractions` and sets them, and then forward to that constructor from the child. Then the child looks like `ParentConstructor(param x, param y) : ChildConstructor(x, y)`

kripken commented on src/passes/Precompute.cpp:
```diff
@@ -39,106 +39,54 @@
+// Limit loop iterations since loops might be infinite.
+static const Index MAX_LOOP_ITERATIONS = 3;
+
 typedef std::unordered_map<LocalGet*, Literals> GetValues;
```
please add a comment explaining how these get values differ from the parent class's.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -150,10 +150,18 @@ class Indenter {
-  Index maxDepth;
+  Index maxDepth = 0; // = no limit
+
+  Index maxLoopIterations = 0; // = no limit
```
this 0 is confusing I think. while a max depth of 0 is meaningless, a max loop iterations of 0 is valid - it means to avoid any loop iterations.

how about using `std::limits<Index>::max()` (i.e., default to the maximum-sized integer)?

dcodeIO replied:
While looking at this I figured that 1 iteration is somewhat analogous to traversing a block (that we give up on if it loops), which is always safe to do and in fact is the only valid preset for precompute since 2+ iterations would require a side effect, which we cannot replace.

Now, if someone wants to disable looking at blocks, they have to override `visitBlock`, and similarly one would have to override `visitLoop` instead of setting `maxLoopIterations=0`. As such I'd somewhat prefer keeping the `0` here, since `max` might be considered a valid limit. Also maps well to `maxDepth` being disabled with `0`. Wdyt?

kripken replied:
I see, fair enough.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -150,10 +150,18 @@ class Indenter {
 
   Index depth = 0;
 
+  Module* module = nullptr;
```
please add a comment saying why this is optional.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1246,167 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+        return visit(curr->value);
+      }
+      // Otherwise remember the constant value set, if any, for subsequent gets.
+      if (setLocalValue(curr->index, curr->value)) {
```
The code around here would be simpler, I think, if we computed `visit(curr->value)` unconditionally first, and then decided what to return later. The code right not hides the computation inside the call to `setLocalValue`, which is less obvious.

Do we even need `setLocalValue(index, expression)`? Can't it always take an index and a value?

dcodeIO replied:
The one place where an expression is useful is in the JS API, where dealing with literals is somewhat an internal mechanism not directly exposed via the API. Moved this to the C API only now, so the parent runner doesn't need the expression variant :)

dcodeIO commented:
What I'm not particularly fond about is how this now (not) affects `RuntimeExpressionRunner`, which even though it extends `ExpressionRunner` has its very own concept of a module and a function scope, leading to the base `module`, `localValues`, `globalValues` and `flags` fields being completely irrelevant. In this particular scenario, the earlier inheritance structure made sense, hmm.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -149,11 +149,51 @@ class Indenter {
+  // Flags indicating special requirements. See FlagValues.
+  Flags flags;
 
+  // Maximum depth before giving up. Defaults to 0 for no limit.
```
It's risky to put stuff like "defaults to 0" in comments, as they tend to become stale.

I think it's best to just assign to 0 here in code, and the same later down.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -149,11 +149,51 @@ class Indenter {
+  // Optional module context to search for globals and called functions. NULL if
+  // we are not interested in any context or if the subclass uses an alternative
+  // mechanism, like RuntimeExpressionRunner does.
+  Module* module;
```
and this should be set to `nullptr`.

kripken commented on src/wasm-interpreter.h:
```diff
@@ -170,11 +210,36 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
-  ExpressionRunner(Index maxDepth) : maxDepth(maxDepth) {}
+  ExpressionRunner(Flags flags = FlagValues::DEFAULT,
+                   Index maxDepth = 0,
+                   Index maxLoopIterations = 0)
```
how about removing the `= 0`s here, so that we have (as suggested earlier) just a single place with `= 0`, earlier. For here we don't need a default value. But let's add a plain `ExpressionRunner() {}` constructor, which then would use those defaults.

kripken replied:
(or maybe a constructor that only takes the flags? depends on what we need. preferably as few constructors as possible, but also without those `= 0`s duplicated)

dcodeIO replied:
Made this three ctors now:

1. A plain one because it's good to have
2. One taking just `maxDepth`, which is what `ConstantExpressionRunner` and `RuntimeExpressionRunner` use
3. One taking all of `module`, `flags`, `maxDepth` and `maxLoopIterations` which is what `Precompute` and the C-API use

plus making the `0` a `NO_LIMIT` constant for clarity in code. That ok? :)

kripken commented on src/wasm-interpreter.h:
```diff
@@ -170,11 +210,36 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+  ExpressionRunner(Module* module,
+                   Flags flags = FlagValues::DEFAULT,
+                   Index maxDepth = 0,
+                   Index maxLoopIterations = 0)
```
also remove `= 0`s here.

(the flag values can remain with a default as the value is not a surprising one. but that means it should move to the end)

kripken commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1243,171 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+        return visit(curr->value);
+      }
+      // Otherwise remember the constant value set, if any, for subsequent gets.
+      auto setFlow = visit(curr->value);
```
this line could be on line 1262, and then the value used instead of having two identical calls in this function.

dcodeIO commented:
First post was missing the new `maxLoopIterations` parameter, but should be up to date now. Biggest point of concern for me is `ConstantExpressionRunner` and `RuntimeExpressionRunner` inheriting the new stuff as well, which both don't use it. I hope nothing bad can happen there.

tlively commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1242,171 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+        if (func->sig.results.isConcrete()) {
+          auto numOperands = curr->operands.size();
+          assert(numOperands == func->getNumParams());
+          auto prevLocalValues = localValues;
```
It looks like this can be a `std::move`, too. Not sure whether that makes the subsequent `localValues.clear()` redundant or not, though.

dcodeIO replied:
Heh, have been thinking about these lines way too much as well, but ultimately decided to stay safe since `indicate that an object t may be "moved from"` doesn't make any guarantees.

tlively commented on src/passes/Precompute.cpp:
```diff
@@ -39,53 +39,44 @@
+      getValues(getValues) {}
 
-  struct NonstandaloneException {
+  struct NonconstantException {
```
Can this be defined in `ExpressionRunner` instead so it doesn't need to be duplicated in all the subclasses?

dcodeIO replied:
Moved it to the parent, and as expected this also helped to solve https://github.com/WebAssembly/binaryen/pull/2702#discussion_r404625808

tlively commented on src/wasm-interpreter.h:
```diff
@@ -2240,6 +2463,22 @@ class ModuleInstance
 };
 
+// Expression runner exposed by the C-API
+class CExpressionRunner final : public ExpressionRunner<CExpressionRunner> {
```
Is there any reason not to define this subclass in binaryen-c.cpp and the PrecomputingExpressionRunner in Precompute.cpp?

dcodeIO replied:
Defining it in `binaryen-c.cpp` yielded an error that the size of `struct NonconstantException` is `0 in C and 1 in C++`. Going to move that struct as per your suggestion, so might well be that this isn't an issue anymore.

dcodeIO replied:
Yep, that worked

tlively commented on test/binaryen.js/expressionrunner.js:
```diff
@@ -0,0 +1,156 @@
+    module.i32.const(1)
+  )
+);
+assert(expr === 0);
```
Would it be more idiomatic for the JS API to turn this into `null` or something like that?

dcodeIO replied:
Figured that one would typically test this as `!expr` anyway, just doing an overly precise check here for testing purposes. Would imagine that not mixing `0` and `null` has benefits for the JIT.

dcodeIO replied:
An alternative here is to return the unmodified original expression. Would that be an improvement?

tlively commented on test/binaryen.js/expressionrunner.js:
```diff
@@ -0,0 +1,156 @@
+    module.local.get(0, binaryen.i32)
+  )
+);
+assert(JSON.stringify(binaryen.getExpressionInfo(expr)) === '{"id":14,"type":2,"value":8}');
```
I think it would make these tests easier to understand if the JSON were not hardcoded with the raw numbers for `id` and `type`. Would it be possible to explicitly construct the expected expressions instead?

dcodeIO replied:
Used the respective constants now and added a little `assertDeepEqual` to make it more easily readable.

tlively commented on test/passes/precompute_all-features.txt:
```diff
@@ -258,7 +258,10 @@
  )
- (func $reftype-test (; 18 ;) (result nullref)
+ (func $loop-precompute (; 18 ;) (result i32)
+  (i32.const 1)
```
Awesome 👍

tlively commented on src/binaryen-c.cpp:
```diff
@@ -4882,6 +4882,98 @@ BinaryenExpressionRef RelooperRenderAndDispose(RelooperRef relooper,
+}
+
+BinaryenExpressionRef
+ExpressionRunnerRunAndDispose(ExpressionRunnerRef runner,
```
If the runner fails, it seems like it would be useful to expose more information to the caller about why it failed. That way the user could choose to increase the depth or loop count, if applicable. What do you think?

dcodeIO replied:
Hmm, good question. Seems like this might be a bit too much, considering how it complicates the API. For instance, on the AssemblyScript side I expect to always use a reasonable `maxDepth` (or none) and give up otherwise as there is no reason to make an exception using larger limits. Would have used that limit right away then.

tlively commented on src/binaryen-c.cpp:
```diff
@@ -4882,6 +4882,98 @@ BinaryenExpressionRef RelooperRenderAndDispose(RelooperRef relooper,
+                                           BinaryenIndex maxDepth,
+                                           BinaryenIndex maxLoopIterations) {
+  if (tracing) {
+    std::cout << "  the_runner = ExpressionRunnerCreate(the_module, " << flags
```
It's unfortunate that this will only work correctly if there is at most one Runner created at a time, but it's probably not worth fixing urgently. Perhaps you could at least leave a TODO about it?

dcodeIO replied:
Implemented something working, but the code for it turned out to be a bit unattractive since, other than expressions etc., runners can be deleted leading to undefined behavior in tracing. Added comments.

dcodeIO commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1242,171 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+    // Check if a constant value has been set in the context of this runner.
+    auto iter = localValues.find(curr->index);
+    if (iter != localValues.end()) {
+      return Flow(std::move(iter->second));
```
My understanding here is that creating the `std::pair<const wasm::Index, wasm::Literals>` will make a copy of the `wasm::Literals` *value*, so using a `std::move` here hints that we can move that volatile copy instead of copying twice and dumping one. Perfectly possible that I don't actually know what I'm doing. Please advise :)

dcodeIO commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1242,171 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+        return Flow();
+      }
+    }
+    return Flow(NONCONSTANT_FLOW);
```
Doing so would duplicate the `return Flow(NONCONSTANT_FLOW)`. That still ok?

dcodeIO commented on src/binaryen-c.h:
```diff
@@ -1642,6 +1642,60 @@ BINARYEN_API void RelooperAddBranchForSwitch(RelooperBlockRef from,
+BINARYEN_API ExpressionRunnerFlags ExpressionRunnerFlagsDefault();
+
+// Be very careful to preserve any side effects, like those of a `local.tee`,
+// for example when we are going to replace the expression afterwards.
```
This now reads

```
// Be very careful to preserve any side effects. For example, if we are
// intending to replace the expression with a constant afterwards, even if we
// can technically evaluate down to a constant, we still cannot replace the
// expression if it also sets a local, which must be preserved in this scenario
// so subsequent code keeps functioning.
```

dcodeIO commented on src/binaryen-c.h:
```diff
@@ -1642,6 +1642,60 @@ BINARYEN_API void RelooperAddBranchForSwitch(RelooperBlockRef from,
+// Traverse through function calls, attempting to compute their concrete value.
+// Must not be used in function-parallel scenarios, where the called function
+// might or might not have been optimized already to something we can traverse
+// successfully, in turn leading to non-deterministic behavior.
```
Used your suggested change and mentioned that traversing another function uses this runner's flags, which implies `PreserveSideEffects`.

dcodeIO commented on src/binaryen-c.h:
```diff
@@ -1642,6 +1642,60 @@ BINARYEN_API void RelooperAddBranchForSwitch(RelooperBlockRef from,
+                       BinaryenIndex maxLoopIterations);
+
+// Sets a known local value to use. Order matters if expressions have side
+// effects. Returns `true` if the expression actually evaluates to a constant.
```
This now reads

```
// Sets a known local value to use. Order matters if expressions have side
// effects. For example, if the expression also sets a local, this side effect
// will also happen (not affected by any flags). Returns `true` if the
// expression actually evaluates to a constant.
```

dcodeIO commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1242,171 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+      }
+      // Otherwise remember the constant value set, if any, for subsequent gets.
+      if (!setFlow.breaking()) {
+        setLocalValue(curr->index, setFlow.values);
```
Good catch, updated the code accordingly

dcodeIO commented on src/wasm-interpreter.h:
```diff
@@ -1174,28 +1242,171 @@ class ExpressionRunner : public OverriddenVisitor<SubType, Flow> {
+    // Check if a constant value has been set in the context of this runner.
+    auto iter = globalValues.find(curr->name);
+    if (iter != globalValues.end()) {
+      return Flow(std::move(iter->second));
```
See https://github.com/WebAssembly/binaryen/pull/2702#discussion_r404613604. My assumptions on how this works might be wrong.

kripken commented:
Is this ready to land, or still waiting for review from @tlively ?

tlively commented:
I'll take a final look now.

tlively commented on src/binaryen-c.cpp:
```diff
@@ -208,6 +209,26 @@ size_t noteExpression(BinaryenExpressionRef expression) {
+
+// Even though unlikely, it is possible that we are trying to use an id that is
+// still in use after wrapping around, which we must prevent.
+std::unordered_set<size_t> usedExpressionRunnerIds;
```
This should probably be `static`, too. You could get extra fancy by making both of these helpers `static` variables inside of `noteExpressionRunner` to limit their scope, but I'll leave that up to you. OTOH, it would probably be better to just say we don't support making more than max size_t expression runners and get rid of all this logic, especially since it is literally impossible to have than many expression runners recorded in `expressionRunners`.

dcodeIO replied:
Mostly thinking in terms of a very long lived process using Binaryen, let's say where modules are being created as-a-service. While we can't store max size_t in the structure, we might at some point overflow, where the likely scenario is that this is just fine, yet guarding for not reusing something left over seems like a good precaution to have. Unlikely that someone will do this with tracing enabled, ofc.

tlively replied:
Aha, I had missed that the `ExpressionRunner`s were removed from the `expressionRunners` map when they were destroyed 👍

tlively commented on test/binaryen.js/expressionrunner.js:
```diff
@@ -5,6 +5,15 @@ console.log("// ExpressionRunner.Flags.TraverseCalls = " + Flags.TraverseCalls);
+function assertDeepEqual(x, y) {
+  if (typeof x === "object") {
+    for (var i in x) assertDeepEqual(x[i], y[i]);
+    for (i in y) assertDeepEqual(x[i], y[i]);
```
Can we do `var i` here, too, or would that be unidiomatic or bad? Seeing the variable reused like this gives me the heebie jeebies.

dcodeIO commented:
On it, will fix these two real quick :)

tlively commented:
@dcodeIO How is this for a commit message? (This just copied from the opening description)

------

Refactors most of the precompute pass's expression runner into its base class so it can also be used via the C and JS APIs. Also adds the option to populate the runner with known constant local and global values upfront, and remembers assigned intermediate values as well as traversing into functions if requested.

C-API:
ExpressionRunnerFlagsDefault()
ExpressionRunnerFlagsPreserveSideeffects()
ExpressionRunnerFlagsTraverseCalls()
ExpressionRunnerCreate(module, flags, maxDepth, maxLoopIterations)
ExpressionRunnerSetLocalValue(runner, index, value)
ExpressionRunnerSetGlobalValue(runner, name, value)
ExpressionRunnerRunAndDispose(runner, expr)

JS-API:
binaryen.ExpressionRunner.Flags.Default
binaryen.ExpressionRunner.Flags.PreserveSideeffects
binaryen.ExpressionRunner.Flags.TraverseCalls
new binaryen.ExpressionRunner(module[, flags[, maxDepth[, maxLoopIterations]]])
binaryen.ExpressionRunner#setLocalValue(index, valueExpr)
binaryen.ExpressionRunner#setGlobalValue(name, valueExpr)
binaryen.ExpressionRunner#runAndDispose(expr)

dcodeIO commented:
Looks good :) (have been trying for a while to keep the first post good for a commit message, sometimes divided by a horizontal line to indicate where additional comments start)

tlively commented:
@dcodeIO Looks like there is a merge conflict to resolve now :(

kripken commented:
Ok great, merging! Thanks for all the work here, and sorry it took this long, but sometimes more complex changes end up that way...

aheejin commented:
I saw this now, and sorry for late questions. It looks this PR duplicates many of functionalities of `RuntimeExpressionRunner` in `ExpressionRunner`. Why is that? I only took a brief look and I might well be mistaken, so please advise!

dcodeIO commented:
Was under the impression that `RuntimeExpressionRunner` is a different beast and requires a lot of context, like memory and external interface, where what I wanted to achieve was to quickly evaluate an expression with limited context while the module is still being generated, for example to check if a static condition (i.e. in generics supporting varying types) is statically `true` or `false`, again affecting codegen. Some of these changes are also beneficial to the precompute pass (or previously lived in a separate PrecomputeExpressionRunner), which does something very similar.

aheejin commented:
Thanks for the answer. Yes, I now think the functionalities duplicated are not in `RuntimeExpressionRunner` but more in `PrecomputeExpressionRunner`. I opened #2797 for this.
